crates/agent/src/lib.rs

```diff
@@ -86,60 +85,50 @@ async fn listen_for_tasks(
     // This will then bubble through as a None return to PgListener:try_recv(),
     // which we'll then attempt to reconnect, and trigger the task handlers
     // to look and see if there's any work waiting for them.
-    let cloned_listener = listener.clone();
-    let handle = tokio::spawn(async move {
-        loop {
-            tokio::time::sleep(Duration::from_secs(30)).await;
-            cloned_listener
-                .lock()
-                .await
-                .listen("hacky_keepalive")
-                .await
-                .unwrap();
-
-            cloned_listener
-                .lock()
-                .await
-                .unlisten("hacky_keepalive")
-                .await
-                .unwrap();
-        }
-    });
-
     loop {
-        // try_recv returns None when the channel disconnects,
-        // which we want to have explicit handling for
-        let maybe_item = listener
-            .lock()
-            .await
-            .try_recv()
-            .await
-            .map_err(|e| anyhow::Error::from(e))?;
-
-        if let Some(item) = maybe_item {
-            let notification: AgentNotification = serde_json::from_str(item.payload())
-                .context("deserializing agent task notification")?;
-
-            tracing::debug!(
-                table = &notification.table,
-                "Message received to invoke handler"
-            );
-            task_tx
-                .send(HandlerInvocation {
-                    table_name: notification.table,
-                    is_poll: false,
-                })
-                .map_err(|e| anyhow::Error::from(e))?
-        } else {
-            tracing::warn!("LISTEN/NOTIFY stream from postgres lost, waking all handlers and attempting to reconnect");
-            table_names.iter().for_each(|table| {
-                task_tx
-                    .send(HandlerInvocation {
-                        table_name: table.clone(),
-                        is_poll: false,
-                    })
-                    .unwrap()
-            });
+        let should_poke_connection: Result<bool, anyhow::Error> = tokio::select! {
```

Comment:
Okay, a few things:

I've universally found it's better to have as little code inside of a select! as possible. It gets wonky from a rust-analyzer & IDE perspective, type hinting doesn't work well, etc. Plus, it's more confusing to read.

* You can extract the tokio timer to a variable.
* How about a mut bool `should_poke` defined _outside_ of the loop, and handling for that being true at the _top_ of the loop?

Doing these to things gets you to:

```
let timer = tokio::time::sleep(Duration::from_secs(30));
// tokio::pin! as needed
let notify = tokio::select! {
  _ = timer => {
    should_poke = true;
    continue;
  },
  notify = listener.try_recv() => notify,
};

// Handle `notify`.

```

I'll further suggest peeling off the error case below as:

```
let Some(item) = maybe_item else {
  tracing::warn!("LISTEN/NOTIFY stream from postgres lost ...
  ... handling ...
  continue;
}

// Now you can handle `item` as more of a through-line for the reader.
```